src/generic/complex_matrices.cc

```diff
@@ -538,6 +539,647 @@ namespace oomph
     }
   }
 
+  //===========================================================================
+  /// Function to multiply this matrix by the CRDoubleMatrix matrix_in.
+  /// In a serial matrix, there are 4 methods available:
+  /// Method 1: First runs through this matrix and matrix_in to find the storage
+  ///           requirements for result - arrays of the correct size are
+  ///           then allocated before performing the calculation.
+  ///           Minimises memory requirements but more costly.
+  /// Method 2: Grows storage for values and column indices of result 'on the
+  ///           fly' using an array of maps. Faster but more memory
+  ///           intensive.
+  /// Method 3: Grows storage for values and column indices of result 'on the
+  ///           fly' using a vector of vectors. Not particularly impressive
+  ///           on the platforms we tried...
+  /// Method 2 is employed by default.
+  //=============================================================================
+  void CRComplexMatrix::multiply(const CRComplexMatrix& matrix_in,
+                                 CRComplexMatrix& result) const
+  {
+#ifdef PARANOID
+    // check that this matrix is built
+    if (!Built)
+    {
+      std::ostringstream error_message_stream;
+      error_message_stream << "This matrix has not been built";
+      throw OomphLibError(error_message_stream.str(),
+                          OOMPH_CURRENT_FUNCTION,
+                          OOMPH_EXCEPTION_LOCATION);
+    }
+    // check that this matrix is built
+    if (!matrix_in.built())
+    {
+      std::ostringstream error_message_stream;
+      error_message_stream << "This matrix matrix_in has not been built";
+      throw OomphLibError(error_message_stream.str(),
+                          OOMPH_CURRENT_FUNCTION,
+                          OOMPH_EXCEPTION_LOCATION);
+    }
+    // if soln is setup then it should have the same distribution as x
+    if (result.built())
+    {
+      if (!(*result.distribution_pt() == *this->distribution_pt()))
+      {
+        std::ostringstream error_message_stream;
+        error_message_stream
+          << "The matrix result is setup and therefore must have the same "
+          << "distribution as the vector x";
+        throw OomphLibError(error_message_stream.str(),
+                            OOMPH_CURRENT_FUNCTION,
+                            OOMPH_EXCEPTION_LOCATION);
+      }
+    }
+#endif
+
+    // short name for Serial_matrix_matrix_multiply_method
+    unsigned method = 2;
```

Comment:
What's everyone's thoughts on implementing 3 methods and using only one?

Reply:
That's slightly odd. I think the value is meant to be exposed so that the user can control the method to use. The implementation of `CRDoubleMatrix::multiply()` sets `method` using a private variable (`Serial_matrix_matrix_multiply_method `):

https://github.com/oomph-lib/oomph-lib/blob/5f89893b2383918b59971310464826609af47f3a/src/generic/matrices.cc#L2036-L2037

that can be set by the user 

https://github.com/oomph-lib/oomph-lib/blob/5f89893b2383918b59971310464826609af47f3a/src/generic/matrices.h#L1159-L1162

I propose that the same approach be taken here.

Reply:
agree! (once the other, possibly long-defunct methods have been checked, possibly permanently via self-test). If not it may be safer to keep only the most efficient one (if that's easy to judge).

Reply:
I haven't tested the other methods and I haven't got round to figuring out the self-tests yet. I would be happier to keep the best/current method and delete the rest. If a speed up/reduced memory requirement is needed, we should be able to isolate and re-implement this later, I guess?
Which would be the best method to keep? Just the current one?

Reply:
There's probably (possibly!) a reason why this one was kept but you should check this with Andrew @alhazel since he was involved in this project. Puneet and I weren't!

Reply:
Implemented a switch and unit self tests for each of the methods using small 4 by 4 matrices. They all pass on my machine.